# SaaS Product Usage  & Customer Retention EDA

### Importing the Libraries

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

#### Loading the Dataset

In [2]:
df = pd.read_csv('saas_business.csv')

In [3]:
df.head(10)

,customer_id,gender,age,country,city,customer_segment,tenure_months,signup_channel,contract_type,monthly_logins,...,avg_resolution_time,complaint_type,csat_score,escalations,email_open_rate,marketing_click_rate,nps_score,survey_response,referral_count,churn
0,CUST_00001,Male,68,Bangladesh,London,SME,22,Web,Monthly,26,...,13.354360,Service,4.0,0,0.71,0.40,27,Satisfied,1,0
1,CUST_00002,Female,57,Canada,Sydney,Individual,9,Mobile,Monthly,7,...,25.140088,Billing,2.0,0,0.78,0.33,-19,Neutral,2,1
2,CUST_00003,Male,24,Germany,New York,SME,58,Web,Yearly,19,...,27.572928,Service,3.0,0,0.35,0.49,80,Neutral,1,0
3,CUST_00004,Male,49,Australia,Dhaka,Individual,19,Mobile,Yearly,34,...,26.420822,Technical,5.0,1,0.83,0.15,100,Neutral,0,0
4,CUST_00005,Male,65,Bangladesh,Delhi,Individual,52,Web,Monthly,20,...,26.674579,Technical,4.0,0,0.65,0.44,21,Unsatisfied,1,0
5,CUST_00006,Female,36,India,London,Individual,54,Mobile,Monthly,1,...,36.038391,Technical,4.0,1,0.83,0.05,-40,Neutral,1,0
6,CUST_00007,Male,23,India,Toronto,SME,54,Mobile,Monthly,21,...,36.892928,Billing,4.0,0,0.75,0.46,-3,Neutral,1,0
7,CUST_00008,Male,37,Bangladesh,Sydney,Individual,7,Referral,Yearly,17,...,25.024133,Technical,4.0,1,0.54,0.13,28,Satisfied,1,0
8,CUST_00009,Male,44,India,Delhi,SME,15,Referral,Monthly,29,...,32.465536,Technical,2.0,0,0.85,0.13,-24,Satisfied,0,0
9,CUST_00010,Female,70,Bangladesh,Berlin,Individual,17,Mobile,Monthly,18,...,42.986192,Billing,4.0,0,0.72,0.13,17,Neutral,1,0


#### Displaying the total number of rows and columns in the dataset

In [4]:
rows, columns = df.shape

print(f'This dataset conatins {rows} rows and {columns} columns.')

This dataset conatins 10000 rows and 32 columns.


#### Displaying column names

In [5]:
df.columns

Index(['customer_id', 'gender', 'age', 'country', 'city', 'customer_segment',
       'tenure_months', 'signup_channel', 'contract_type', 'monthly_logins',
       'weekly_active_days', 'avg_session_time', 'features_used',
       'usage_growth_rate', 'last_login_days_ago', 'monthly_fee',
       'total_revenue', 'payment_method', 'payment_failures',
       'discount_applied', 'price_increase_last_3m', 'support_tickets',
       'avg_resolution_time', 'complaint_type', 'csat_score', 'escalations',
       'email_open_rate', 'marketing_click_rate', 'nps_score',
       'survey_response', 'referral_count', 'churn'],
      dtype='str')

#### Based on the dataset description provided on Kaggle, the meaning of each column in the dataset is as follows:

- customer_id: Unique identifier assigned to each customer.
- gender: Gender of the customer (e.g., Male, Female, Other).
- age: Age of the customer in years.
- country: Country where the customer is located.
- city: City where the customer resides.
- customer_segment: Business-defined customer classification (e.g., SMB, Enterprise, Individual).
- tenure_months: Number of months the customer has been subscribed to the service.
- signup_channel: Channel through which the customer signed up (e.g., Organic, Ads, Referral).
- contract_type: Type of subscription contract (e.g., Monthly, Quarterly, Annual).

Product Usage Metrics
- monthly_logins: Total number of times the customer logged into the platform in a month.
- weekly_active_days: Average number of days per week the customer actively used the platform.
- avg_session_time: Average duration of each user session (typically in minutes).
- features_used: Number of distinct product features used by the customer.
- usage_growth_rate: Percentage change in usage over a defined recent period.
- last_login_days_ago: Number of days since the customer last logged in.

Billing & Payment Information
- monthly_fee: Subscription fee charged to the customer per month.
- total_revenue: Total revenue generated from the customer over their lifetime.
- payment_method: Method used for payment (e.g., Credit Card, PayPal, Bank Transfer).
- payment_failures: Number of failed payment attempts recorded.
- discount_applied: Indicates whether a discount has been applied to the customer’s subscription (Yes/No or binary flag).
- price_increase_last_3m: Indicates whether the customer experienced a price increase in the last three months (Yes/No or binary flag).

Customer Support & Satisfaction
- support_tickets: Number of customer support tickets raised.
- avg_resolution_time: Average time taken to resolve support tickets (e.g., in hours or days).
- complaint_type: Category of complaint raised by the customer (e.g., Billing, Technical, Service).
- csat_score: Customer Satisfaction Score (typically on a defined scale such as 1–5).
- escalations: Number of support cases escalated beyond first-level support.

Engagement & Feedback
- email_open_rate: Percentage of marketing or communication emails opened by the customer.
- marketing_click_rate: Percentage of marketing emails where links were clicked.
- nps_score: Net Promoter Score provided by the customer (typically ranging from -100 to 100 or 0 to 10 depending on scale).
- survey_response: Indicates whether the customer responded to feedback surveys (Yes/No or binary flag).
- referral_count: Number of new customers referred by the existing customer.

Critical Variable
- churn: Binary target variable indicating whether the customer discontinued the service (0 = No, 1 = Yes).

In [6]:
df.describe()

,age,tenure_months,monthly_logins,weekly_active_days,avg_session_time,features_used,usage_growth_rate,last_login_days_ago,monthly_fee,total_revenue,payment_failures,support_tickets,avg_resolution_time,csat_score,escalations,email_open_rate,marketing_click_rate,nps_score,referral_count,churn
count,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.0000,10000.000000
mean,45.906100,30.155700,19.672000,3.479400,15.187890,4.990800,0.019423,9.505200,34.930000,1057.018000,0.501000,1.206600,23.953210,3.486700,0.294100,0.499610,0.253575,19.110500,0.9922,0.102100
std,16.420416,17.099517,9.838003,2.302283,6.830774,2.211607,0.149567,9.799417,23.785115,1020.148233,0.711934,1.104554,9.956999,0.978631,0.543171,0.231601,0.140532,38.936619,0.9939,0.302795
min,18.000000,1.000000,0.000000,0.000000,1.000000,1.000000,-0.580000,0.000000,10.000000,10.000000,0.000000,0.000000,1.000000,1.000000,0.000000,0.100000,0.010000,-100.000000,0.0000,0.000000
25%,32.000000,16.000000,13.000000,1.000000,10.391645,3.000000,-0.080000,2.000000,20.000000,340.000000,0.000000,0.000000,17.082041,3.000000,0.000000,0.300000,0.130000,-7.000000,0.0000,0.000000
50%,46.000000,30.000000,20.000000,3.000000,15.161659,5.000000,0.020000,6.000000,30.000000,720.000000,0.000000,1.000000,23.950444,4.000000,0.000000,0.500000,0.250000,19.000000,1.0000,0.000000
75%,60.000000,45.000000,26.000000,5.000000,19.882034,6.000000,0.120000,13.000000,50.000000,1440.000000,1.000000,2.000000,30.758480,4.000000,1.000000,0.700000,0.380000,46.000000,2.0000,0.000000
max,74.000000,59.000000,54.000000,7.000000,41.996411,15.000000,0.540000,80.000000,100.000000,5900.000000,5.000000,7.000000,61.821878,5.000000,4.000000,0.900000,0.500000,100.000000,7.0000,1.000000


#### Displaying columns along with their data types

In [7]:
df.dtypes

customer_id                   str
gender                        str
age                         int64
country                       str
city                          str
customer_segment              str
tenure_months               int64
signup_channel                str
contract_type                 str
monthly_logins              int64
weekly_active_days          int64
avg_session_time          float64
features_used               int64
usage_growth_rate         float64
last_login_days_ago         int64
monthly_fee                 int64
total_revenue               int64
payment_method                str
payment_failures            int64
discount_applied              str
price_increase_last_3m        str
support_tickets             int64
avg_resolution_time       float64
complaint_type                str
csat_score                float64
escalations                 int64
email_open_rate           float64
marketing_click_rate      float64
nps_score                   int64
survey_respons

#### Observations on Data Types

- Clearly, most columns in the dataset are stored in appropriate data types for analysis. However, a few adjustments would improve analytical clarity, memory efficiency, and modeling readiness.

- For example, churn is stored as int64, but it represents a binary categorical outcome (0 = No, 1 = Yes). While keeping it as integer is acceptable for machine learning models, converting it to a categorical type is more appropriate for exploratory data analysis and visualization.

- Similarly, nps_score is stored as int64. Although numeric, it often behaves like an ordinal or categorical metric in business analysis (e.g., Detractors, Passives, Promoters). Depending on the analysis goal, it may require grouping or transformation.

- Additionally, several categorical variables such as gender, country, city, customer_segment, signup_channel, contract_type, payment_method, complaint_type, discount_applied, price_increase_last_3m, and survey_response are stored as object (string) types. While technically correct, converting them to the category data type would improve memory usage and analytical performance.

#### Checking all unique values of each column

In [8]:
for col in df.columns:
    unique_vals = df[col].unique()
    print(f'Column Name : {col}')
    print(f'Total number of unique values : {len(unique_vals)}')
    print(f'Unique Values : {unique_vals}')
    print('-'*50)
    print('\n')

Column Name : customer_id
Total number of unique values : 10000
Unique Values : <StringArray>
['CUST_00001', 'CUST_00002', 'CUST_00003', 'CUST_00004', 'CUST_00005',
 'CUST_00006', 'CUST_00007', 'CUST_00008', 'CUST_00009', 'CUST_00010',
 ...
 'CUST_09991', 'CUST_09992', 'CUST_09993', 'CUST_09994', 'CUST_09995',
 'CUST_09996', 'CUST_09997', 'CUST_09998', 'CUST_09999', 'CUST_10000']
Length: 10000, dtype: str
--------------------------------------------------


Column Name : gender
Total number of unique values : 2
Unique Values : <StringArray>
['Male', 'Female']
Length: 2, dtype: str
--------------------------------------------------


Column Name : age
Total number of unique values : 57
Unique Values : [68 57 24 49 65 36 23 37 44 70 60 54 22 28 32 20 40 26 46 34 43 38 25 74
 55 59 41 52 72 31 27 29 61 63 51 58 48 18 56 21 30 39 35 19 53 71 67 47
 73 50 45 69 42 64 66 33 62]
--------------------------------------------------


Column Name : country
Total number of unique values : 7
Unique

#### Checking for any missing values per column

In [9]:
# Count of NaN / None values per column
nan_count = df.isnull().sum()

# Count of empty strings per column
empty_count = (df == '').sum()

# Total missing values (NaN + empty strings)
total_missing = nan_count + empty_count

# Percentage of missing values
missing_percent = (total_missing / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Values': total_missing,
    'Percentage (%)': missing_percent
}).sort_values(by='Percentage (%)', ascending=False)

print(missing_df)

                        Missing Values  Percentage (%)
complaint_type                    2045           20.45
customer_id                          0            0.00
age                                  0            0.00
gender                               0            0.00
city                                 0            0.00
customer_segment                     0            0.00
tenure_months                        0            0.00
country                              0            0.00
contract_type                        0            0.00
monthly_logins                       0            0.00
weekly_active_days                   0            0.00
avg_session_time                     0            0.00
features_used                        0            0.00
usage_growth_rate                    0            0.00
last_login_days_ago                  0            0.00
signup_channel                       0            0.00
monthly_fee                          0            0.00
total_reve

#### Summary from the Observations

Columns with Missing Values
- complaint_type - 2,045 missing values (20.45%) 


For 'complaint_type', replacing NaN with the mode is NOT recommended, because NaN here has a business meaning: it indicates the customer did NOT make any complaint.

Imputing with the most common complaint type would distort the data and bias analysis.

Instead, we replace NaN with a meaningful category, e.g., "No Complaint", which preserves the business interpretation and allows easy analysis and modeling.


In [10]:
df['complaint_type'] = df['complaint_type'].fillna('No Complaint')

#### nps_score

We’re converting the aggregated NPS score into ordinal categories to simplify analysis and easily compare customer sentiment with churn patterns.

- Detractor: overall negative sentiment
- Passive: neutral or moderate sentiment
- Promoter: highly positive sentiment

In [11]:
def nps_ordinal(nps):
    if nps > 50:
        return 'Promoter'
    elif nps >= 0:
        return 'Passive'
    else:  # nps < 0
        return 'Detractor'

df['nps_category'] = df['nps_score'].apply(nps_ordinal)

# To Check counts
print(df['nps_category'].value_counts())

nps_category
Passive      4827
Detractor    3009
Promoter     2164
Name: count, dtype: int64


#### Converting the String Columns to 'category type'

In [12]:
# Convert object/string columns to categorical type to improve memory usage and performance
categorical_cols = [
    'gender', 'country', 'city', 'customer_segment', 'signup_channel',
    'contract_type', 'payment_method', 'complaint_type', 'discount_applied',
    'price_increase_last_3m', 'survey_response', 'nps_category'                   
]

for col in categorical_cols:
    df[col] = df[col].astype('category')

In [13]:
df.dtypes

customer_id                    str
gender                    category
age                          int64
country                   category
city                      category
customer_segment          category
tenure_months                int64
signup_channel            category
contract_type             category
monthly_logins               int64
weekly_active_days           int64
avg_session_time           float64
features_used                int64
usage_growth_rate          float64
last_login_days_ago          int64
monthly_fee                  int64
total_revenue                int64
payment_method            category
payment_failures             int64
discount_applied          category
price_increase_last_3m    category
support_tickets              int64
avg_resolution_time        float64
complaint_type            category
csat_score                 float64
escalations                  int64
email_open_rate            float64
marketing_click_rate       float64
nps_score           

## Time for Analysis

#### Univaraite Analysis

In [14]:
# -------------------------
# CATEGORICAL VARIABLES
# -------------------------
categorical_cols = [
    'gender',
    'country',
    'customer_segment',
    'signup_channel',
    'contract_type',
    'payment_method',
    'discount_applied',
    'price_increase_last_3m',
    'complaint_type',
    'survey_response',
    'nps_category'
]

for col in categorical_cols:
    counts = (
        df[col]
        .value_counts()
        .rename_axis(col)
        .reset_index(name='Count')
    )

    fig = px.bar(
        counts,
        x='Count',
        y=col,
        orientation='h',
        text='Count',
        title=f'{col.replace("_"," ").title()} Distribution'
    )

    fig.update_layout(yaxis={'categoryorder':'total descending'})
    fig.update_traces(textposition='outside')
    fig.show()


In [15]:
# -------------------------
# NUMERIC VARIABLES
# -------------------------

numeric_cols = [
    'age',
    'tenure_months',
    'monthly_logins',
    'weekly_active_days',
    'avg_session_time',
    'features_used',
    'usage_growth_rate',
    'last_login_days_ago',
    'monthly_fee',
    'total_revenue',
    'payment_failures',
    'support_tickets',
    'avg_resolution_time',
    'csat_score',
    'nps_score',
    'email_open_rate',
    'marketing_click_rate',
    'referral_count'
]

for col in numeric_cols:
    fig = px.histogram(
        df,
        x=col,
        nbins=30,
        title=f'Distribution of {col.replace("_"," ").title()}',
        labels={col: col.replace("_"," ").title(), 'count':'Count'}
    )

    fig.update_layout(
        xaxis_title=col.replace("_"," ").title(),
        yaxis_title='Count',
        bargap=0.1
    )

    fig.show()


In [16]:
# -------------------------
# BOOLEAN VARIABLES
# -------------------------

df['churn_label'] = df['churn'].map({0: 'Retained', 1: 'Churned'})

churn_counts = (
    df['churn_label']
    .value_counts()
    .rename_axis('Churn Status')
    .reset_index(name='Count')
)

fig = px.bar(
    churn_counts,
    x='Churn Status',
    y='Count',
    text='Count',
    title='Churn Distribution'
)

fig.update_traces(textposition='outside')
fig.show()


#### Bivaraite Analysis

In [17]:
import plotly.express as px

# ===================================================== 
# # NUMERIC vs CATEGORICAL (Mostly vs Churn) 
# =====================================================

numeric_cols = [
    'tenure_months',
    'monthly_logins',
    'features_used',
    'usage_growth_rate',
    'last_login_days_ago',
    'support_tickets',
    'payment_failures',
    'csat_score',
    'nps_score',
    'total_revenue'
]

for col in numeric_cols:
    fig = px.histogram(
        df,
        x=col,
        color='churn_label',
        barmode='overlay',   # overlay makes distributions easier to compare
        opacity=0.6,
        nbins=30,
        title=f'{col.replace("_"," ").title()} Distribution by Churn',
        labels={col: col.replace("_"," ").title(), 'churn_label':'Churn Status'}
    )
    fig.show()


In [18]:
# =====================================================
# CATEGORICAL vs CATEGORICAL (Stacked Bars)
# =====================================================

categorical_vs_churn = [
    'contract_type',
    'customer_segment',
    'signup_channel',
    'payment_method',
    'discount_applied',
    'price_increase_last_3m',
    'complaint_type',
    'survey_response',
    'nps_category'
]

for col in categorical_vs_churn:
    counts = (
        df.groupby([col, 'churn_label'])
        .size()
        .reset_index(name='Count')
    )

    fig = px.bar(
        counts,
        x=col,
        y='Count',
        color='churn_label',
        barmode='stack',
        title=f'{col.replace("_"," ").title()} vs Churn',
        text='Count'
    )

    fig.update_traces(textposition='outside')
    fig.show()

In [19]:
# =====================================================
# NUMERIC vs NUMERIC (Colored by Churn)
# =====================================================

# 1 Monthly Logins vs Features Used
fig = px.scatter(
    df,
    x='monthly_logins',
    y='features_used',
    color='churn_label',
    title='Monthly Logins vs Features Used by Churn',
    labels={
        'monthly_logins': 'Monthly Logins',
        'features_used': 'Features Used',
        'churn_label': 'Churn Status'
    }
)
fig.show()


# 2 Support Tickets vs CSAT
fig = px.scatter(
    df,
    x='support_tickets',
    y='csat_score',
    color='churn_label',
    size='escalations',
    title='Support Tickets vs CSAT (Bubble Size = Escalations)',
    labels={
        'support_tickets': 'Support Tickets',
        'csat_score': 'CSAT Score',
        'churn_label': 'Churn Status'
    }
)
fig.show()


# 3 Usage Growth Rate vs NPS
fig = px.scatter(
    df,
    x='usage_growth_rate',
    y='nps_score',
    color='churn_label',
    title='Usage Growth Rate vs NPS Score by Churn',
    labels={
        'usage_growth_rate': 'Usage Growth Rate',
        'nps_score': 'NPS Score',
        'churn_label': 'Churn Status'
    }
)
fig.show()

In [20]:
# =====================================================
# LIFECYCLE ANALYSIS (Tenure Bucket vs Churn)
# =====================================================

df['tenure_bucket'] = pd.cut(
    df['tenure_months'],
    bins=[0,3,6,12,24,60],
    labels=['0-3','3-6','6-12','12-24','24+']
)

counts = (
    df.groupby(['tenure_bucket', 'churn_label'])
    .size()
    .reset_index(name='Count')
)

fig = px.bar(
    counts,
    x='tenure_bucket',
    y='Count',
    color='churn_label',
    barmode='stack',
    title='Tenure Bucket vs Churn',
    labels={
        'tenure_bucket': 'Tenure Bucket (Months)',
        'churn_label': 'Churn Status'
    }
)

fig.show()

In [21]:
# Email Open Rate vs Marketing Click Rate by Churn
fig = px.scatter(
    df,
    x='email_open_rate',
    y='marketing_click_rate',
    color='churn_label',
    title='Email Open Rate vs Marketing Click Rate by Churn',
    labels={
        'email_open_rate': 'Email Open Rate',
        'marketing_click_rate': 'Marketing Click Rate',
        'churn_label': 'Churn Status'
    }
)

fig.show()

In [22]:
# Referral Count Distribution by Churn
fig = px.histogram(
    df,
    x='referral_count',
    color='churn_label',
    barmode='overlay',   # Overlay both churn groups
    nbins=20,
    opacity=0.6,
    title='Referral Count Distribution by Churn',
    labels={
        'referral_count': 'Referral Count',
        'churn_label': 'Churn Status'
    }
)

fig.show()

#### Multivaraite Analysis

In [23]:
# Ensuring churn label
df['churn_label'] = df['churn'].map({0:'Retained', 1:'Churned'})

# -------------------------
# 1. CSAT by Contract Type and Customer Segment (Grouped Boxplot)
# -------------------------
fig = px.box(
    df,
    x='contract_type',
    y='csat_score',
    color='customer_segment',
    title='CSAT Score by Contract Type and Customer Segment',
    labels={'contract_type':'Contract Type','csat_score':'CSAT Score','customer_segment':'Customer Segment'}
)

fig.update_layout(
    title_font_size=18,
    xaxis_title_font_size=14,
    yaxis_title_font_size=14,
    xaxis_tickfont_size=12,
    yaxis_tickfont_size=12,
    legend=dict(title_font_size=14, font_size=12)
)
fig.show()

# -------------------------
# 2. Features Used × Usage Growth Rate × Churn (Scatter Heatmap)
# -------------------------
scatter_data = df.groupby(
    ['features_used','usage_growth_rate','churn_label']
).size().reset_index(name='Count')

fig = px.density_heatmap(
    scatter_data,
    x='features_used',
    y='usage_growth_rate',
    z='Count',
    facet_col='churn_label',
    title='Features Used × Usage Growth Rate × Churn',
    labels={'Count':'Number of Users','features_used':'Features Used','usage_growth_rate':'Usage Growth Rate'}
)

fig.update_layout(
    title_font_size=18,
    xaxis_title_font_size=14,
    yaxis_title_font_size=14,
    xaxis_tickfont_size=10,
    yaxis_tickfont_size=12,
    legend=dict(title_font_size=14, font_size=12)
)
fig.show()

# -------------------------
# 3. Support Tickets × Avg Resolution Time × Churn (Faceted Scatter)
# -------------------------
fig = px.scatter(
    df,
    x='support_tickets',
    y='avg_resolution_time',
    color='churn_label',
    facet_col='contract_type',
    size='escalations',   # optional bubble size
    opacity=0.6,
    title='Support Tickets vs Avg Resolution Time by Contract Type & Churn',
    labels={'support_tickets':'Support Tickets','avg_resolution_time':'Avg Resolution Time','churn_label':'Churn Status'}
)
fig.show()

# -------------------------
# 4. Monthly Logins × Features Used × Contract Type (Faceted Scatter)
# -------------------------
fig = px.scatter(
    df,
    x='monthly_logins',
    y='features_used',
    color='churn_label',
    facet_col='contract_type',
    opacity=0.6,
    title='Monthly Logins vs Features Used by Contract Type & Churn',
    labels={'monthly_logins':'Monthly Logins','features_used':'Features Used','churn_label':'Churn Status'}
)
fig.show()

# -------------------------
# 5. NPS Category × Churn × Customer Segment (Heatmap)
# -------------------------
heatmap_data = df.groupby(['nps_category','customer_segment','churn_label']).size().reset_index(name='Count')

fig = px.density_heatmap(
    heatmap_data,
    x='nps_category',
    y='customer_segment',
    z='Count',
    facet_col='churn_label',
    title='NPS Category × Customer Segment × Churn',
    labels={'Count':'Number of Users','nps_category':'NPS Category','customer_segment':'Customer Segment'}
)

fig.update_layout(
    width=1400,
    height=400,
    title_font_size=18,
    xaxis_title_font_size=14,
    yaxis_title_font_size=14,
    xaxis_tickfont_size=10,
    yaxis_tickfont_size=12,
    legend=dict(title_font_size=14, font_size=12)
)
fig.show()


The following graphs are recreations of the above graphs and are included solely to provide images for the report.

In [24]:
# -------------------------
# Function to style scatter plots
# -------------------------
def style_scatter(fig, xaxis_title, yaxis_title, title_text):
    
    fig.update_traces(
        marker=dict(size=10, opacity=0.7),
        selector=dict(mode='markers')
    )
    
    fig.update_layout(
        width=1500,
        height=800,
        xaxis_title=dict(text=xaxis_title, font=dict(size=25, color='black')),
        yaxis_title=dict(text=yaxis_title, font=dict(size=25, color='black')),
        xaxis=dict(showgrid=False, tickfont=dict(size=20, color='black')),
        yaxis=dict(showgrid=False, tickfont=dict(size=20, color='black')),
        title=dict(
            text=title_text,
            font=dict(size=25, color='black'),
            x=0.5,
            xanchor='center'
        ),
        plot_bgcolor='rgba(0,0,0,0)',
        paper_bgcolor='rgba(0,0,0,0)',
        legend=dict(
            font=dict(size=18, color='black')
        )
    )
    
    return fig


# -------------------------
# 1 Monthly Logins vs Features Used
# -------------------------
fig = px.scatter(
    df,
    x='monthly_logins',
    y='features_used',
    color='churn_label',
    color_discrete_sequence=['#1982c4', '#ff6b6b'],  # consistent clean palette
    labels={
        'monthly_logins': 'Monthly Logins',
        'features_used': 'Features Used',
        'churn_label': 'Churn Status'
    }
)

fig = style_scatter(
    fig,
    'Monthly Logins',
    'Features Used',
    'Monthly Logins vs Features Used by Churn'
)

fig.show()


In [ ]:
# -------------------------
# Function to style faceted scatter plots
# -------------------------
def style_facet_scatter(fig, xaxis_title, yaxis_title, title_text):

    fig.update_traces(
        marker=dict(
            sizemode='area',
            line=dict(width=1, color='black')
        ),
        selector=dict(mode='markers')
    )

    # Apply consistent axis styling to ALL facet axes
    fig.for_each_xaxis(lambda axis: axis.update(
        showgrid=False,
        tickfont=dict(size=20, color='black'),
        title_font=dict(size=25, color='black')
    ))

    fig.for_each_yaxis(lambda axis: axis.update(
        showgrid=False,
        tickfont=dict(size=20, color='black'),
        title_font=dict(size=25, color='black')
    ))

    # Style facet titles (Monthly, Yearly, Quarterly)
    fig.for_each_annotation(
        lambda a: a.update(
            text=a.text.split("=")[-1],
            font=dict(size=22, color='black')
        )
    )

    fig.update_layout(
        width=1500,
        height=800,
        title=dict(
            text=title_text,
            font=dict(size=25, color='black'),
            x=0.5,
            xanchor='center'
        ),
        plot_bgcolor='rgba(0,0,0,0)',
        paper_bgcolor='rgba(0,0,0,0)',
        legend=dict(font=dict(size=18, color='black'))
    )

    return fig




# -------------------------
# Support Tickets × Avg Resolution Time × Churn
# -------------------------
fig = px.scatter(
    df,
    x='support_tickets',
    y='avg_resolution_time',
    color='churn_label',
    facet_col='contract_type',
    size='escalations',
    opacity=0.6,
    color_discrete_sequence=['#1982c4', '#ff6b6b'],
    labels={
        'support_tickets':'Support Tickets',
        'avg_resolution_time':'Avg Resolution Time',
        'churn_label':'Churn Status'
    }
)

fig = style_facet_scatter(
    fig,
    'Support Tickets',
    'Avg Resolution Time',
    'Support Tickets vs Avg Resolution Time by Contract Type & Churn'
)

fig.show()


In [ ]:
# -------------------------
# NPS Score Distribution by Churn
# -------------------------
fig = px.histogram(
    df,
    x='nps_score',
    color='churn_label',
    barmode='overlay',
    opacity=0.6,
    nbins=30,
    color_discrete_sequence=['#1982c4', '#ff6b6b'],
    labels={
        'nps_score': 'NPS Score',
        'churn_label': 'Churn Status'
    }
)

# Apply clean styling
fig.update_layout(
    width=1500,
    height=800,
    xaxis_title=dict(text='NPS Score', font=dict(size=25, color='black')),
    yaxis_title=dict(text='Number of Customers', font=dict(size=25, color='black')),
    xaxis=dict(showgrid=False, tickfont=dict(size=20, color='black')),
    yaxis=dict(showgrid=False, tickfont=dict(size=20, color='black')),
    title=dict(
        text='NPS Score Distribution by Churn',
        font=dict(size=25, color='black'),
        x=0.5,
        xanchor='center'
    ),
    plot_bgcolor='rgba(0,0,0,0)',
    paper_bgcolor='rgba(0,0,0,0)',
    legend=dict(font=dict(size=18, color='black'))
)

fig.show()


In [ ]:
# -------------------------
# Email Open Rate vs Marketing Click Rate by Churn
# -------------------------
fig = px.scatter(
    df,
    x='email_open_rate',
    y='marketing_click_rate',
    color='churn_label',
    opacity=0.6,
    color_discrete_sequence=['#1982c4', '#ff6b6b'],
    labels={
        'email_open_rate': 'Email Open Rate',
        'marketing_click_rate': 'Marketing Click Rate',
        'churn_label': 'Churn Status'
    }
)

# Improve marker appearance
fig.update_traces(
    marker=dict(
        size=10,
        line=dict(width=1, color='black')
    )
)

# Apply clean layout styling
fig.update_layout(
    width=1500,
    height=800,
    xaxis_title=dict(text='Email Open Rate', font=dict(size=25, color='black')),
    yaxis_title=dict(text='Marketing Click Rate', font=dict(size=25, color='black')),
    xaxis=dict(showgrid=False, tickfont=dict(size=20, color='black')),
    yaxis=dict(showgrid=False, tickfont=dict(size=20, color='black')),
    title=dict(
        text='Email Open Rate vs Marketing Click Rate by Churn',
        font=dict(size=25, color='black'),
        x=0.5,
        xanchor='center'
    ),
    plot_bgcolor='rgba(0,0,0,0)',
    paper_bgcolor='rgba(0,0,0,0)',
    legend=dict(font=dict(size=18, color='black'))
)

fig.show()